forte/processors/base/data_augment_processor.py

```diff
@@ -0,0 +1,92 @@
+# Copyright 2020 The Forte Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Processors that augment the data. The processor will call
+replacement ops to generate texts similar to those in the input pack
+and create a new pack with them.
+"""
+from typing import List, Tuple
+from forte.data.ontology.core import Entry
+from forte.processors.base.base_processor import BaseProcessor
+from forte.processors.data_augment.algorithms.text_replacement_op \
+    import TextReplacementOp
+
+__all__ = [
+    "BaseDataAugmentProcessor",
+    "ReplacementDataAugmentProcessor"
+]
+
+
+class BaseDataAugmentProcessor(BaseProcessor):
+    r"""The base class of processors that augment data.
+    This processor instantiates replacement ops where specific
+    data augmentation algorithms are implemented. The replacement ops
+    will run the algorithms and the processor will create Forte
+    data structures based on the augmented inputs.
+    """
+
+
+class ReplacementDataAugmentProcessor(BaseDataAugmentProcessor):
+    r"""
+    Most of the Data Augmentation(DA) methods can be
+    considered as replacement-based methods with different
+    levels: character, word, sentence or document.
+    """
+    def __init__(self):
+        """
+        The replaced_spans records the entries replaced by new texts.
+        It consists of tuples (entry, new text) inserted by :func: replace.
+        The new text will be used for building new data pack.
+        """
+        super().__init__()
+        self.replaced_spans: List[Tuple[Entry, str]] = []
+
+    def replace(self, replacement_op: TextReplacementOp, input: Entry):
+        """
+        This is a wrapper function to call the replacement op. After
+        getting the augmented text, it will register the input & output
+        for later batch process of building the new data pack.
+        """
+        replaced_text: str = replacement_op.replace(input)
+        self.replaced_spans.append((input, replaced_text))
```

Comment:
Now maybe we can use this information to do the alignment. It will be great if this processor calls the function automatically so that the implementation can focus on the algorithm logic.

By the way, we have a somewhat similar functionality implemented [here](https://github.com/asyml/forte/blob/master/forte/data/data_pack.py#L180), which is used when the users remove "noisy text" from the original data (i.e. HTML tags). Maybe it can help you with this function.

Reply:
The batch replacement function(I am still working on) may be called in the _process function of the processor. I was also thinking about testing the base classes. I guess we can test it after finishing the batch replacement function. But maybe I will create another PR for that function, so that we can merge the base classes ASAP.